gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -262,25 +309,35 @@ public Duration getStreamIdleTimeout() {
     return streamIdleTimeout;
   }
 
-  /**
-   * The channel affinity for this context.
-   *
-   * @see ApiCallContext#withStreamIdleTimeout(Duration)
-   */
-  @BetaApi("The surface for streaming is not stable yet and may change in the future.")
+  /** The channel affinity for this context. */
+  @BetaApi("The surface for channel affinity is not stable yet and may change in the future.")
   @Nullable
   public Integer getChannelAffinity() {
     return channelAffinity;
   }
 
+  /** The extra header for this context. */
+  @BetaApi("The surface for extra headers is not stable yet and may change in the future.")
+  @Override
+  public Map<String, List<String>> getExtraHeaders() {
+    ImmutableMap.Builder<String, List<String>> builder = ImmutableMap.builder();
```

Comment:
I am not quite sure if we want to make the returned `Map` immutable. This [link](https://docs.oracle.com/javase/7/docs/api/java/net/URLConnection.html#getHeaderFields%28%29) that Garrett mentioned made the returned map unmodifiedable, but I myself don't have a strong argument.

Reply:
Since `String` is also immutable, I believe returning immutable map is more than appropriate here (the only correct solution). It should be Immutable map with immutable keys (strings) and immutable lists of strings as values. If Map<String, List<String>> was used as internal representation, it would be possible just remove an internal list here (no any conversions), since the outer code would not be able to damage the  immutable returned map in any way.